File: archive/original_repo/Sages_Stone-main/systems/RCF/RCF.go
// rcf.go - Reality Constraint Fuzzer
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type State struct {
	Vector []float64
	Intent float64
	Memory float64
}

func NewState(vector []float64, intent, memory float64) *State {
	return &State{Vector: vector, Intent: intent, Memory: memory}
}

func (s *State) Magnitude() float64 {
	sum := 0.0
	for _, v := range s.Vector {
		sum += v * v
	}
	return math.Sqrt(sum)
}

func (s *State) Clone() *State {
	vec := make([]float64, len(s.Vector))
	copy(vec, s.Vector)
	return &State{Vector: vec, Intent: s.Intent, Memory: s.Memory}
}

type Constraint struct {
	Name           string
	Check          func(float64) bool
	Persona        string
	Adaptive       bool
	ViolationCount int
}

func NewConstraint(name string, check func(float64) bool, persona string, adaptive bool) *Constraint {
	return &Constraint{
		Name:           name,
		Check:          check,
		Persona:        persona,
		Adaptive:       adaptive,
		ViolationCount: 0,
	}
}

func (c *Constraint) Evaluate(value float64) bool {
	ok := c.Check(value)
	if !ok {
		c.ViolationCount++
	}
	return ok
}

func (c *Constraint) Voice() string {
	voices := map[string]string{
		"Judge":     "The Stern Judge: 'Order must be kept.'",
		"Trickster": "The Trickster: 'Boundaries are illusions.'",
		"Guardian":  "The Guardian: 'I protect the fragile edge.'",
		"Neutral":   "Constraint speaks without drama.",
	}
	if v, ok := voices[c.Persona]; ok {
		return v
	}
	return voices["Neutral"]
}

type HardenedNet struct {
	Constraints []*Constraint
	Limit       float64
	FatigueMap  map[int]int
}

func NewHardenedNet(constraints []*Constraint, limit float64) *HardenedNet {
	return &HardenedNet{
		Constraints: constraints,
		Limit:       limit,
		FatigueMap:  make(map[int]int),
	}
}

func (n *HardenedNet) Tension(value, drift float64) (int, string) {
	for _, c := range n.Constraints {
		if !c.Evaluate(value) {
			return -1, fmt.Sprintf("Collision → %s", c.Voice())
		}
	}

	sector := int(math.Round(value * 10))
	n.FatigueMap[sector]++

	if n.FatigueMap[sector] > 5 && math.Abs(value-math.Round(value)) < 0.05 {
		return -2, "Crystallization → 'The familiar became brittle.'"
	}

	return 1, ""
}

type ContractingNet struct {
	Base         *HardenedNet
	GlobalStress float64
}

func NewContractingNet(constraints []*Constraint, limit float64) *ContractingNet {
	return &ContractingNet{
		Base:         NewHardenedNet(constraints, limit),
		GlobalStress: 0.0,
	}
}

func (n *ContractingNet) Tension(value, drift float64) (int, string) {
	n.GlobalStress += 0.02 + math.Abs(drift) + 0.01
	effectiveLimit := n.Base.Limit - n.GlobalStress

	if math.Abs(value) >= effectiveLimit || effectiveLimit <= 0.1 {
		return -3, "Global Collapse → 'The envelope shrank to silence.'"
	}

	return n.Base.Tension(value, drift)
}

// Spike Functions
func staticFn(s *State) *State {
	vec := make([]float64, len(s.Vector))
	for i, v := range s.Vector {
		vec[i] = v * 0.9
	}
	return NewState(vec, s.Intent+0.1, s.Memory)
}

func entropyWeatherFn(s *State, storm bool) *State {
	scale := 0.02
	if storm {
		scale = 0.08
	}

	vec := make([]float64, len(s.Vector))
	for i, v := range s.Vector {
		vec[i] = v + rand.NormFloat64()*scale
	}

	intentDelta := 0.1
	if storm {
		intentDelta = 0.3
	}

	return NewState(vec, s.Intent+intentDelta, s.Memory)
}

func resonanceFn(s *State) *State {
	factor := 1.01 + 0.01*math.Sin(s.Intent)
	vec := make([]float64, len(s.Vector))
	for i, v := range s.Vector {
		vec[i] = v * factor
	}
	return NewState(vec, s.Intent+0.5, s.Memory)
}

func anchorFn(s *State) *State {
	vec := make([]float64, len(s.Vector))
	for i, v := range s.Vector {
		vec[i] = v * 0.5
	}
	return NewState(vec, s.Intent-1.0, s.Memory)
}

func memoryFn(s *State) *State {
	vec := make([]float64, len(s.Vector))
	copy(vec, s.Vector)
	return NewState(vec, s.Intent, s.Memory+0.2)
}

func intentPressureFn(s *State) *State {
	vec := make([]float64, len(s.Vector))
	for i, v := range s.Vector {
		vec[i] = v + s.Intent*0.01
	}
	return NewState(vec, s.Intent, s.Memory)
}

type TransformationEngine struct{}

func (e *TransformationEngine) Transform(s *State) (*State, float64) {
	sum := 0.0
	for _, v := range s.Vector {
		sum += v
	}
	mean := sum / float64(len(s.Vector))
	drift := (0.5 - math.Mod(mean, 1.0)) * 0.015

	vec := make([]float64, len(s.Vector))
	for i, v := range s.Vector {
		vec[i] = v + drift
	}

	return NewState(vec, s.Intent, s.Memory), drift
}

type SimulationND struct {
	State      *State
	Net        *ContractingNet
	Engine     *TransformationEngine
	Frames     int
	History    map[string][]float64
	PoeticLog  []string
	Survivors  []*State
}

func NewSimulationND(dimensions int, constraints []*Constraint, limit float64, frames int) *SimulationND {
	vec := make([]float64, dimensions)
	for i := range vec {
		vec[i] = rand.Float64()*1.0 + 0.5
	}

	return &SimulationND{
		State:  NewState(vec, 0.0, 0.0),
		Net:    NewContractingNet(constraints, limit),
		Engine: &TransformationEngine{},
		Frames: frames,
		History: map[string][]float64{
			"volley":    {},
			"magnitude": {},
			"intent":    {},
			"memory":    {},
		},
		PoeticLog: []string{},
		Survivors: []*State{},
	}
}

func (sim *SimulationND) Step(volley int) (int, string) {
	storm := volley%20 == 0

	sim.State = staticFn(sim.State)
	sim.State = entropyWeatherFn(sim.State, storm)

	transformed, drift := sim.Engine.Transform(sim.State)
	sim.State = transformed

	if volley%4 == 0 {
		sim.State = resonanceFn(sim.State)
	}
	if volley > 30 {
		sim.State = anchorFn(sim.State)
	}
	if volley%5 == 0 {
		sim.State = memoryFn(sim.State)
	}
	if volley%7 == 0 {
		sim.State = intentPressureFn(sim.State)
	}

	code, message := sim.Net.Tension(sim.State.Magnitude(), drift)

	sim.History["volley"] = append(sim.History["volley"], float64(volley))
	sim.History["magnitude"] = append(sim.History["magnitude"], sim.State.Magnitude())
	sim.History["intent"] = append(sim.History["intent"], sim.State.Intent)
	sim.History["memory"] = append(sim.History["memory"], sim.State.Memory)

	if message != "" {
		sim.PoeticLog = append(sim.PoeticLog, message)
	}

	return code, message
}

func (sim *SimulationND) Run() {
	collapsed := false

	for volley := 0; volley < sim.Frames; volley++ {
		code, message := sim.Step(volley)

		if code < 0 {
			fmt.Printf("[!] Terminated at volley %d: %s\n", volley, message)
			collapsed = true
			break
		}
	}

	if !collapsed {
		fmt.Println("✓ Stable run achieved – 'The dance continued without collapse.'")
		sim.Survivors = append(sim.Survivors, sim.State.Clone())
	}

	fmt.Println("\n--- Emergent Narrative ---")
	for _, line := range sim.PoeticLog {
		fmt.Println(line)
	}

	if len(sim.Survivors) > 0 {
		fmt.Println("\n--- Survivor Genealogy ---")
		for i, s := range sim.Survivors {
			fmt.Printf("Ancestor %d: Magnitude=%.3f, Intent=%.2f, Memory=%.2f\n",
				i+1, s.Magnitude(), s.Intent, s.Memory)
		}
	}

	fmt.Println("\n--- Commentary ---")
	if len(sim.PoeticLog) > 0 {
		fmt.Println("Alan Watts voice: 'Collapse was not failure, but revelation. The system discovered itself.'")
	} else {
		fmt.Println("Alan Watts voice: 'No collapse today. Stability is also discovery.'")
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())

	fmt.Println(strings.Repeat("=", 60))
	fmt.Println("Reality Constraint Fuzzer (Polyform Engine)")
	fmt.Println(strings.Repeat("=", 60))

	constraints := []*Constraint{
		NewConstraint("HardBounds", func(x float64) bool {
			return math.Abs(x) > 0.1 && math.Abs(x) < 5.0
		}, "Judge", false),
		NewConstraint("AntiInteger", func(x float64) bool {
			return math.Abs(x-math.Round(x)) > 0.05
		}, "Trickster", false),
	}

	sim := NewSimulationND(3, constraints, 5.0, 200)
	sim.Run()
}

File: archive/original_repo/Sages_Stone-main/systems/RCF/RCF.js
// RCF.js - Reality Constraint Fuzzer
// Node.js/Browser compatible

class State {
  constructor(vector, intent = 0.0, memory = 0.0) {
    this.vector = vector;
    this.intent = intent;
    this.memory = memory;
  }

  magnitude() {
    return Math.sqrt(this.vector.reduce((sum, v) => sum + v * v, 0));
  }

  clone() {
    return new State([...this.vector], this.intent, this.memory);
  }
}

class Constraint {
  constructor(name, check, persona = "Neutral", adaptive = false) {
    this.name = name;
    this.check = check;
    this.persona = persona;
    this.adaptive = adaptive;
    this.violationCount = 0;
  }

  evaluate(value) {
    const ok = this.check(value);
    if (!ok) this.violationCount++;
    return ok;
  }

  adapt() {
    if (this.adaptive && this.violationCount > 3) {
      this.check = (x) => Math.abs(x) > 0.1 && Math.abs(x) < 4.5;
    }
  }

  voice() {
    const voices = {
      Judge: "The Stern Judge: 'Order must be kept.'",
      Trickster: "The Trickster: 'Boundaries are illusions.'",
      Guardian: "The Guardian: 'I protect the fragile edge.'",
      Neutral: "Constraint speaks without drama."
    };
    return voices[this.persona] || voices.Neutral;
  }
}

class HardenedNet {
  constructor(constraints, limit = 5.0) {
    this.constraints = constraints;
    this.limit = limit;
    this.fatigueMap = new Map();
  }

  tension(value, drift) {
    // Check constraints
    for (const c of this.constraints) {
      if (!c.evaluate(value)) {
        return { code: -1, message: `Collision → ${c.voice()}` };
      }
    }

    // Check fatigue
    const sector = Math.round(value * 10) / 10;
    const fatigue = this.fatigueMap.get(sector) || 0;
    this.fatigueMap.set(sector, fatigue + 1);

    if (fatigue > 5 && Math.abs(value - Math.round(value)) < 0.05) {
      return { code: -2, message: "Crystallization → 'The familiar became brittle.'" };
    }

    return { code: 1, message: null };
  }
}

class ContractingNet extends HardenedNet {
  constructor(constraints, limit = 5.0) {
    super(constraints, limit);
    this.globalStress = 0.0;
  }

  tension(value, drift) {
    this.globalStress += 0.02 + Math.abs(drift) + 0.01;
    const effectiveLimit = this.limit - this.globalStress;

    if (Math.abs(value) >= effectiveLimit || effectiveLimit <= 0.1) {
      return { code: -3, message: "Global Collapse → 'The envelope shrank to silence.'" };
    }

    return super.tension(value, drift);
  }
}

// Spike Functions
function staticFn(s) {
  return new State(
    s.vector.map(v => v * 0.9),
    s.intent + 0.1,
    s.memory
  );
}

function entropyWeatherFn(s, storm = false) {
  const scale = storm ? 0.08 : 0.02;
  return new State(
    s.vector.map(v => v + (Math.random() - 0.5) * 2 * scale),
    s.intent + (storm ? 0.3 : 0.1),
    s.memory
  );
}

function resonanceFn(s) {
  const factor = 1.01 + 0.01 * Math.sin(s.intent);
  return new State(
    s.vector.map(v => v * factor),
    s.intent + 0.5,
    s.memory
  );
}

function anchorFn(s) {
  return new State(
    s.vector.map(v => v * 0.5),
    s.intent - 1.0,
    s.memory
  );
}

function memoryFn(s) {
  return new State([...s.vector], s.intent, s.memory + 0.2);
}

function intentPressureFn(s) {
  return new State(
    s.vector.map(v => v + s.intent * 0.01),
    s.intent,
    s.memory
  );
}

class TransformationEngine {
  transform(s) {
    const mean = s.vector.reduce((a, b) => a + b, 0) / s.vector.length;
    const drift = (0.5 - (mean % 1)) * 0.015;
    return {
      state: new State(
        s.vector.map(v => v + drift),
        s.intent,
        s.memory
      ),
      drift
    };
  }
}

class SimulationND {
  constructor(dimensions = 3, constraints = [], limit = 5.0, frames = 200) {
    this.dimensions = dimensions;
    this.constraints = constraints;
    this.limit = limit;
    this.frames = frames;
    
    this.state = new State(
      Array.from({ length: dimensions }, () => Math.random() * 1.0 + 0.5),
      0.0,
      0.0
    );
    
    this.net = new ContractingNet(constraints, limit);
    this.engine = new TransformationEngine();
    
    this.history = {
      volley: [],
      magnitude: [],
      intent: [],
      memory: []
    };
    
    this.poeticLog = [];
    this.survivors = [];
  }

  step(volley) {
    const storm = volley % 20 === 0;
    
    this.state = staticFn(this.state);
    this.state = entropyWeatherFn(this.state, storm);
    
    const { state: transformed, drift } = this.engine.transform(this.state);
    this.state = transformed;
    
    if (volley % 4 === 0) this.state = resonanceFn(this.state);
    if (volley > 30) this.state = anchorFn(this.state);
    if (volley % 5 === 0) this.state = memoryFn(this.state);
    if (volley % 7 === 0) this.state = intentPressureFn(this.state);

    const result = this.net.tension(this.state.magnitude(), drift);
    
    this.history.volley.push(volley);
    this.history.magnitude.push(this.state.magnitude());
    this.history.intent.push(this.state.intent);
    this.history.memory.push(this.state.memory);

    if (result.message) {
      this.poeticLog.push(result.message);
    }

    return result;
  }

  run() {
    let collapsed = false;
    
    for (let volley = 0; volley < this.frames; volley++) {
      const result = this.step(volley);
      
      if (result.code < 0) {
        console.log(`[!] Terminated at volley ${volley}: ${result.message}`);
        collapsed = true;
        break;
      }
    }

    if (!collapsed) {
      console.log("✓ Stable run achieved – 'The dance continued without collapse.'");
      this.survivors.push(this.state.clone());
    }

    console.log("\n--- Emergent Narrative ---");
    this.poeticLog.forEach(line => console.log(line));

    if (this.survivors.length > 0) {
      console.log("\n--- Survivor Genealogy ---");
      this.survivors.forEach((s, i) => {
        console.log(`Ancestor ${i + 1}: Magnitude=${s.magnitude().toFixed(3)}, Intent=${s.intent.toFixed(2)}, Memory=${s.memory.toFixed(2)}`);
      });
    }

    console.log("\n--- Commentary ---");
    if (this.poeticLog.length > 0) {
      console.log("Alan Watts voice: 'Collapse was not failure, but revelation. The system discovered itself.'");
    } else {
      console.log("Alan Watts voice: 'No collapse today. Stability is also discovery.'");
    }

    return this.history;
  }
}

// Example usage
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { State, Constraint, SimulationND };
}

// Run if executed directly
if (typeof require !== 'undefined' && require.main === module) {
  console.log("=".repeat(60));
  console.log("Reality Constraint Fuzzer (Polyform Engine)");
  console.log("=".repeat(60));

  const constraints = [
    new Constraint("HardBounds", x => Math.abs(x) > 0.1 && Math.abs(x) < 5.0, "Judge"),
    new Constraint("AntiInteger", x => Math.abs(x - Math.round(x)) > 0.05, "Trickster")
  ];

  const sim = new SimulationND(3, constraints, 5.0, 200);
  sim.run();
}